src/zenml/orchestrators/utils.py

```diff
@@ -102,10 +103,279 @@ def is_setting_enabled(
+    """Store step outputs in the serve tap for in-memory handoff."""
+    current_tap = _serve_output_tap.get({})
+    current_tap[step_name] = outputs
+    _serve_output_tap.set(current_tap)
+
+
+def tap_get_step_outputs(step_name: str) -> Optional[Dict[str, Any]]:
+    """Get step outputs from the serve tap."""
+    return _serve_output_tap.get({}).get(step_name)
+
+
+def tap_clear() -> None:
+    """Clear the serve tap for a fresh request."""
+    _serve_output_tap.set({})
+
+
+# Response tap for capturing pipeline outputs in-process
+_response_tap: ContextVar[Dict[str, Any]] = ContextVar(
+    "response_tap", default={}
+)
+_return_targets: ContextVar[Dict[str, Optional[str]]] = ContextVar(
+    "return_targets", default={}
+)
+
+
+def response_tap_set(output_name: str, value: Any) -> None:
+    """Set a response output value in the tap.
+
+    Args:
+        output_name: Name of the output in the response
+        value: The output value to store
+    """
+    current_tap = _response_tap.get({})
+    current_tap[output_name] = value
+    _response_tap.set(current_tap)
+
+
+def response_tap_get_all() -> Dict[str, Any]:
+    """Get all captured response outputs.
+
+    Returns:
+        Dictionary of captured outputs
+    """
+    return _response_tap.get({})
+
+
+def response_tap_clear() -> None:
+    """Clear the response tap for a fresh request."""
+    _response_tap.set({})
+
+
+def set_return_targets(targets: Dict[str, Optional[str]]) -> None:
+    """Set the return targets for this request.
+
+    Args:
+        targets: Mapping of step_name -> expected_output_name (or None for first)
+    """
+    _return_targets.set(targets)
+
+
+def get_return_targets() -> Dict[str, Optional[str]]:
+    """Get the return targets for this request.
+
+    Returns:
+        Mapping of step_name -> expected_output_name (or None for first)
+    """
+    return _return_targets.get({})
+
+
+def extract_return_contract(
+    pipeline_source: Optional[str],
+) -> Optional[Dict[str, str]]:
+    """Extract which step outputs to return from pipeline function using AST parsing.
+
+    Args:
+        pipeline_source: Source path of pipeline function
+
+    Returns:
+        Dict mapping output names to step names, or None if extraction fails
+    """
+    if not pipeline_source:
+        return None
+
+    try:
+        import ast
+        import inspect
+
+        from zenml.utils import source_utils
+
+        # Load pipeline function
+        pipeline_func = source_utils.load(pipeline_source)
+        source_code = inspect.getsource(pipeline_func)
+
+        # Parse source code into AST
+        tree = ast.parse(source_code)
+
+        # Find the function definition
+        func_def = None
+        for node in ast.walk(tree):
+            if isinstance(node, ast.FunctionDef):
+                func_def = node
+                break
+
+        if not func_def:
+            return None
+
+        # Find return statements and extract the last one
```

Comment:
Why the last one, if anything only the first one makes sense as the other ones will never be executed. As long as it's a top-level return and not nested inside another inline function

Reply:
outdated